src/com/xilinx/rapidwright/edif/EDIFTokenizer.java

```diff
@@ -0,0 +1,486 @@
+        while (startOffset < endOffset) {
+            int actuallyRead = in.read(buffer, startOffset, endOffset-startOffset);
+            if (actuallyRead == -1) {
+                sawEOF = true;
+                buffer[startOffset] = 0;
+                return false;
+            }
+            available += actuallyRead;
+            startOffset += actuallyRead;
+        }
+        return true;
+    }
+
+    /**
+     * Load more data from stream.
+     *
+     * Only do anything if there is less than one max token length of data available. Then, fill the whole buffer
+     * except one byte. That one byte is set to 0. This way, we never have to re-fill inside long tokens. We also
+     * can detect too-long tokens that are longer than our buffer.
+     */
+    protected void fill() throws IOException {
+        if (available>maxTokenLength || sawEOF) {
+            return;
+        }
+
+        int fillStart = (offset+available)&bufferAddressMask;
+        int fillEnd = (offset-1) & bufferAddressMask;
+
+        buffer[fillEnd] = 0;
+
+        if (fillStart>fillEnd) {
+            //Fill in two parts
+            if (!ensureRead(fillStart, buffer.length)) {
+                return;
+            }
+            ensureRead(0, fillEnd);
+        } else {
+            ensureRead(fillStart, fillEnd);
+        }
+    }
+
+    public EDIFTokenizer(Path fileName, InputStream in, NameUniquifier uniquifier, int maxTokenLength) {
+        this.fileName = fileName;
+        this.in = in;
+        this.uniquifier = uniquifier;
+        this.maxTokenLength = maxTokenLength;
+        //Only a power of two does not share any bits with its lower neighbour
+        if ((maxTokenLength & (maxTokenLength-1)) != 0) {
+            throw new IllegalStateException("max token length must be a power of two but is "+maxTokenLength);
+        }
+        bufferAddressMask = maxTokenLength*2-1;
+        this.buffer = new byte[maxTokenLength*2];
+    }
+
+    public EDIFTokenizer(Path fileName, InputStream in, NameUniquifier uniquifier) {
+        this(fileName, in, uniquifier, DEFAULT_MAX_TOKEN_LENGTH);
+    }
+
+
+    /**
+     * Read two separate locations from a buffer, concatenating them into a single string.
+     * Suppports multi-byte characters split between the two parts
+     * @param buffer the buffer to read from
+     * @param start1 first part start
+     * @param length1 first part length
+     * @param start2 second part start
+     * @param length2 second part length
+     * @return the string assembled from the two locations
+     */
+    public static String byteArrayToStringMulti(byte[] buffer, int start1, int length1, int start2, int length2) {
+        //To support multi-byte characters being split between the parts, we have to take
+        // care to first concatenate, then decode.
+        //TODO benchmark which approach is faster. Arraycopy is probably faster for small strings?
+        if (length1+length2 < 10000) {
```

Comment:
Perhaps label the magic number?

Reply:
I've now benchmarked it. The streams approach is always slower. So no need for the magic number anymore.